### PR TITLE
Soft Delete Stories

### PR DESCRIPTION
**Pull request by andrewyguo**
## Notion ticket link

[Ticket ](https://www.notion.so/uwblueprintexecs/a658c933d18c48d4b43fd99bda2bd3b4?v=ba22ac93d09040ff8219c60ac7caabdb&p=69f5f0a450324aad84df236d4336d1f3)

## Implementation description

- Create new StoryAll and StoryContentAll model to add new column `is_deleted`
- Create new view for Story and StoryContent 
- Backend Changes: 
- Create mutation `SoftDeleteStory`
- Frontend: 
- Call new mutation in `ManageStoryPage`

## Steps to test

1. Login as admin 
2. Go to story page and delete story 
3. Ensure story and story translations are no longer present and you are redirected to the homepage 

## What should reviewers focus on?

- The soft delete story flow makes sense 
- All related StoryContent, StoryTranslations are soft-deleted as well 

## Checklist

- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] For backend changes, I have run the appropriate linters: `docker exec -it planet-read_py-backend_1 /bin/bash -c "black . && isort --profile black ."` and I have generated new migrations: `flask db migrate -m "<your message>"`
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: backend/python/app/services/implementations/story_service.py [hunk 1]
```diff
@@ -216,7 +218,7 @@ def get_story_translations(
                 .order_by(Story.id)
                 .all()
             )
-
+            print("stories: ", stories)
```
Comment (gaoxk) [inline backend/python/app/services/implementations/story_service.py:221]: nit: remove

File: backend/python/app/services/implementations/story_service.py [hunk 2]
```diff
@@ -342,6 +344,7 @@ def get_story_translation_tests(
 
     def get_story_translation(self, id):
         try:
+            print("inside get_story_translation")
```
Comment (gaoxk) [inline backend/python/app/services/implementations/story_service.py:347]: remove

File: backend/python/app/services/implementations/story_service.py [hunk 3]
```diff
@@ -894,6 +897,30 @@ def get_story_translation_statistics(self):
...
         )
 
+    def soft_delete_story(self, id):
+        try:
+            story = StoryAll.query.get(id)
+            for content in story.contents:
+                content.is_deleted = True
+
+            story.is_deleted = True
+            story.translated_languages = None
```
Comment (gaoxk) [inline backend/python/app/services/implementations/story_service.py:907]: remove this line, its not necessary to empty the array and we'd like to preserve information of the stories

File: backend/python/app/services/implementations/story_service.py [hunk 4]
```diff
@@ -894,6 +897,30 @@ def get_story_translation_statistics(self):
...
+
+            # Remove story_translations for this story
```
Comment (gaoxk) [inline backend/python/app/services/implementations/story_service.py:909]: use the service function soft_delete_story_translation on all the story_translations' ids. Soft deleting a story translation involves more than just setting the is_delete col on the story translation

File: backend/python/tools/insert_test_data.py [hunk 5]
```diff
@@ -8,8 +8,12 @@
 db = SQLAlchemy()
 
 from app.models.comment_all import CommentAll
-from app.models.story import Story
-from app.models.story_content import StoryContent
+
+# from app.models.story import Story
+from app.models.story_all import StoryAll
+
+# from app.models.story_content import StoryContent
```
Comment (gaoxk) [inline backend/python/tools/insert_test_data.py:15]: remove comment

File: backend/python/tools/insert_test_data.py [hunk 6]
```diff
@@ -8,8 +8,12 @@
 db = SQLAlchemy()
 
 from app.models.comment_all import CommentAll
-from app.models.story import Story
-from app.models.story_content import StoryContent
+
+# from app.models.story import Story
```
Comment (gaoxk) [inline backend/python/tools/insert_test_data.py:12]: remove comment

File: frontend/src/components/pages/ManageStoryPage.tsx [hunk 7]
```diff
@@ -238,7 +258,7 @@ const ManageStoryPage = () => {
         <ConfirmationModal
           confirmation={confirmDeleteStory}
           onClose={closeModal}
-          onConfirmationClick={() => {}} // TODO: implement story deletion
+          onConfirmationClick={() => callSoftDeleteStoryMutation()} // TODO: implement story deletion
```
Comment (gaoxk) [inline frontend/src/components/pages/ManageStoryPage.tsx:261]: nit: remove TODO comment 😉

**Review by gaoxk — CHANGES_REQUESTED** [review]

File: backend/python/app/tests/graphql/mutations/test_story_mutation.py [hunk 8]
```diff
@@ -1,7 +1,7 @@
 from datetime import datetime
```
Comment (gaoxk) [inline backend/python/app/tests/graphql/mutations/test_story_mutation.py:1]: do you have to use StoryAll and StoryContentAll to pass these tests?

File: backend/python/app/tests/graphql/queries/test_story_query.py [hunk 9]
```diff
@@ -1,5 +1,7 @@
 from ....models.story import Story
```
Comment (gaoxk) [inline backend/python/app/tests/graphql/queries/test_story_query.py:1]: ditto for this

File: backend/python/app/tests/services/test_story_service.py [hunk 10]
```diff
@@ -1,7 +1,9 @@
 import pytest
```
Comment (gaoxk) [inline backend/python/app/tests/services/test_story_service.py:1]: ditto

File: backend/python/tools/insert_test_data.py [hunk 11]
```diff
@@ -36,9 +36,9 @@ def insert_test_data():
     )
 
     # stories
-    db.engine.execute("ALTER TABLE stories AUTO_INCREMENT = 1;")
+    db.engine.execute("ALTER TABLE stories_all AUTO_INCREMENT = 1;")
     db.engine.execute(
-        "INSERT IGNORE INTO stories \
+        "INSERT IGNORE INTO stories_all \
```
Comment (gaoxk) [inline backend/python/tools/insert_test_data.py:41]: i dont think this line change is necessary

File: backend/python/tools/insert_test_data.py [hunk 12]
```diff
@@ -98,12 +98,12 @@ def insert_test_data():
...
 
     # story contents
-    db.engine.execute("ALTER TABLE story_contents AUTO_INCREMENT = 1;")
+    db.engine.execute("ALTER TABLE story_contents_all AUTO_INCREMENT = 1;")
     for story_id in range(1, 8):
         for i, content in enumerate(generic_content):
             id = (story_id - 1) * 10 + i + 1
             db.engine.execute(
-                f'INSERT IGNORE INTO story_contents \
+                f'INSERT IGNORE INTO story_contents_all \
```
Comment (gaoxk) [inline backend/python/tools/insert_test_data.py:106]: i dont think this line change is necessary

File: backend/python/tools/insert_test_data.py [hunk 13]
```diff
@@ -121,7 +121,7 @@ def insert_test_data():
         for i, content in enumerate(story_content):
             id = (story_id - 1) * 10 + i + 1
             db.engine.execute(
-                f'INSERT IGNORE INTO story_contents \
+                f'INSERT IGNORE INTO story_contents_all \
```
Comment (gaoxk) [inline backend/python/tools/insert_test_data.py:124]: i dont think this line change is necessary

File: backend/python/tools/insert_test_data.py [hunk 14]
```diff
@@ -184,7 +184,7 @@ def insert_test_data():
     # Carl (id: 1) and Dwight (id: 4)"
     carl = User.query.filter_by(first_name="Carl").first()
     dwight = User.query.filter_by(first_name="Dwight").first()
-    the_gg = Story.query.filter_by(title="The Great Gatsby").first()
+    the_gg = StoryAll.query.filter_by(title="The Great Gatsby").first()
```
Comment (gaoxk) [inline backend/python/tools/insert_test_data.py:187]: i dont think this line change is necessary

File: frontend/src/components/pages/ManageStoryPage.tsx [hunk 15]
```diff
@@ -109,6 +115,20 @@ const ManageStoryPage = () => {
...
+    try {
+      history.push("/");
+      window.location.reload();
+      await softDeleteStory({
+        variables: {
+          id: parseInt(storyIdParam, 10),
+        },
+      });
+    } catch (error) {
+      window.alert(error ?? "Error occurred, please try again.");
```
Comment (gaoxk) [inline frontend/src/components/pages/ManageStoryPage.tsx:128]: can you include the error in the alert text?

File: backend/python/app/services/implementations/story_service.py [hunk 16]
```diff
@@ -836,9 +837,10 @@ def soft_delete_story_translation(self, id):
             story_translation.is_deleted = True
 
             # Remove language from story.translated_languages
-            story = Story.query.filter_by(id=story_translation.story_id).first()
+            story = StoryAll.query.filter_by(id=story_translation.story_id).first()
```
Comment (gaoxk) [inline backend/python/app/services/implementations/story_service.py:840]: is this change necessary

**Review by gaoxk — CHANGES_REQUESTED** [review]
nice! everything functioned as expected. just have some questions

File: backend/python/app/tests/graphql/mutations/test_story_mutation.py (hunk 8 above)
Comment (andrewyguo) [reply backend/python/app/tests/graphql/mutations/test_story_mutation.py:1]: For this yes. When I revert the changes, `test_create_story` fails

File: backend/python/app/tests/services/test_story_service.py (hunk 10 above)
Comment (andrewyguo) [reply backend/python/app/tests/services/test_story_service.py:1]: Yes. when I revert changes, `test_create_story` fails

File: backend/python/app/tests/graphql/queries/test_story_query.py (hunk 9 above)
Comment (andrewyguo) [reply backend/python/app/tests/graphql/queries/test_story_query.py:1]: Yeah, otherwise the test fails

**Review by gaoxk — APPROVED** [review]
noice 😎
